### DISCUSSION TITLE
본인이 생각하는 좋은 테스트 코드는 어떤 특징이 있나요?

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#42] [Test]
.

**Comment by jimi567** [comment]
저는 최대한 독립적이고, 가독성 및 신뢰성이 높은 테스트가 좋은 테스트라고 생각합니다.
독립적인 테스트가 좋다고 생각하는 근거로는 테스트 간의 의존성을 최소화하여, 각 테스트가 서로 영향을 미치지 않도록 해야 하나의 테스트 실패가 다른 테스트에 영향을 미치지 않기 때문입니다.

가독성의 경우는 테스트 코드는 명세의 역할도 중요한 역할 중 하나라고 생각하기 때문입니다. 테스트의 의도(어떤 도메인 로직에 대한 테스트 인지 등)가 명확해야되고, 어떤 케이스에 대한 테스트인지 등을 명확하게 해두어야 협업을 하는 개발자에게 의도를 명확하게 전달할 수 있기 때문입니다.

또, 리팩토링 과정에서 개발자가 실수를 빠르게 캐치하고, 피드백할 수 있어야 하기 때문에 특정 조건에 있어서 항상 같은 결과가 나오도록 즉, 신뢰성이 높은 테스트를 작성하도록 해야 됩니다.( <- 개인적으로 이게 제일 중요하다고 생각합니다..)

**Comment by nayonsoso** [comment]
첫째로, 명세의 기능할 수 있을 정도로 친절히 작성되어야 합니다.
어떤 기능을 테스트하는지, 테스트에 사용하는 값이 어떤 의미를 갖는지 한눈에 알 수 있어야 합니다.
테스트 코드의 관례인 give-when-then을 단위로 개행을 해주면,
when 부분에서 어떤 함수가 호출되는지를 통해 이를 명확히 할 수 있을 것 같습니다.

둘째로, 테스트는 빨라야 합니다.
이를 위해서 테스트 데이터를 경계값으로 주거나, 중복 테스트를 제거해줘야 합니다.

셋째로, 테스트는 깨지기 쉬워야 합니다.
테스트가 깨지기 쉬워야 어떤 부분에서 사이드 이펙트가 발생했는지를 파악하기 쉽다고 생각합니다.

**Comment by le2sky** [comment]
`잘 설명되는 실패` -> 무엇때문에 실패했는가? 
`이해할 수 있는 테스트 코드` -> 다른 개발자가 봤을 때, 테스트 코드가 정확히 무엇을 테스트하기 위한 것인지 이해 가능?
`쉽고 빠르게 실행` -> 단위 테스트가 느리거나 실행하기 어려우면.. 테스트 실행하기 싫어지고, 개발 시간 낭비
`세부 구현 사항에 독립적` -> 리팩터링 수행 시 테스트 코드를 수정해야 한다면, 신뢰할 수 있는가?
`훼손의 정확한 감지` -> 코드가 훼손되면, 테스트가 반드시 실패해야 함(코드가 훼손된 경우에만 실패해야 함)
`독립적 `-> 다른 테스트에 의존하거나 영향을 주어선 안됨
`반복 가능` -> 몇 번을 다시 진행해도 똑같은 결과가 나와야 함. 랜덤값에 의존적인 테스트를 떠올려볼 것
`철저하고 적절한 때에 작성`-> 나중에 테스트 코드 작성할려면.. 힘들걸?

**Comment by GIVEN53** [comment]
독립적 -> 테스트 간 영향 x
멱등성 -> 몇 번을 실행하든 항상 동일한 결과
가독성 -> 테스트도 문서, 의도 전달
